flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/nn/QuadTree.scala

```diff
@@ -0,0 +1,301 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.ml.nn.util
+
+import org.apache.flink.ml.math.{Breeze, Vector}
+import Breeze._
+
+import org.apache.flink.ml.metrics.distances.DistanceMetric
+
+import scala.collection.mutable.ListBuffer
+import scala.collection.mutable.PriorityQueue
+
+/**
+ * n-dimensional QuadTree data structure; partitions
+ * spatial data for faster queries (e.g. KNN query)
+ * The skeleton of the data structure was initially
+ * based off of the 2D Quadtree found here:
+ * http://www.cs.trinity.edu/~mlewis/CSCI1321-F11/Code/src/util/Quadtree.scala
+ *
+ * Many additional methods were added to the class both for
+ * efficient KNN queries and generalizing to n-dim.
+ *
+ * @param minVec
+ * @param maxVec
+ */
+class QuadTree(minVec:Vector, maxVec:Vector,distMetric:DistanceMetric){
+  var maxPerBox = 20
+
+  class Node(center:Vector,width:Vector, var children:Seq[Node]) {
+
+    var objects = new ListBuffer[Vector]
+
+    /** for testing purposes only; used in QuadTreeSuite.scala
+      *
+      * @return
+      */
+    def getCenterWidth(): (Vector, Vector) = {
+      (center, width)
+    }
+
+    def contains(obj: Vector): Boolean = {
+      overlap(obj, 0.0)
+    }
+
+    /** Tests if obj is within a radius of the node
+      *
+      * @param obj
+      * @param radius
+      * @return
+      */
+    def overlap(obj: Vector, radius: Double): Boolean = {
+      var count = 0
+      for (i <- 0 to obj.size - 1) {
+        if (obj(i) - radius < center(i) + width(i) / 2 &&
+          obj(i) + radius > center(i) - width(i) / 2) {
+          count += 1
+        }
+      }
+
+      if (count == obj.size) {
+        true
+      } else {
+        false
+      }
+    }
+
+    /** Tests if obj is near a node:  minDist is defined so that every point in the box
+      * has distance to obj greater than minDist
+      * (minDist adopted from "Nearest Neighbors Queries" by N. Roussopoulos et al.)
+      *
+      * @param obj
+      * @param radius
+      * @return
+      */
+    def isNear(obj: Vector, radius: Double): Boolean = {
+      if (minDist(obj) < radius) {
+        true
+      } else {
+        false
+      }
+    }
+
+    def minDist(obj: Vector): Double = {
+      var minDist = 0.0
+      for (i <- 0 to obj.size - 1) {
+        if (obj(i) < center(i) - width(i) / 2) {
+          minDist += math.pow(obj(i) - center(i) + width(i) / 2, 2)
+        } else if (obj(i) > center(i) + width(i) / 2) {
+          minDist += math.pow(obj(i) - center(i) - width(i) / 2, 2)
+        }
+      }
+      minDist
+    }
+
+    def whichChild(obj: Vector): Int = {
```

Comment:
Added ScalaDocs, computes the right index for the child when searching through the tree -- the children are indexed, so instead of checking which box by looking at the coordinate values I compute the index.